src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,217 @@
+            return SyntaxFactory.UsingDirective(alias: null, name: name);
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
+
+        private static string GetAliasQualifierOpt(NameSyntax name)
+        {                          
+            while (true)
+            {
+                switch (name.Kind())
+                {
+                    case SyntaxKind.QualifiedName:
+                        name = ((QualifiedNameSyntax)name).Left;
+                        continue;
+                    case SyntaxKind.AliasQualifiedName:
+                        return ((AliasQualifiedNameSyntax)name).Alias.Identifier.ValueText;
+                }
+
+                return null;
+            }
+        }
+        
+        private NameSyntax CreateNameSyntax(ImmutableArray<string> namespaceParts, string aliasQualifier, int index)
+        {
+            var part = namespaceParts[index].EscapeIdentifier();
+            var namePiece = SyntaxFactory.IdentifierName(part);
+
+            if (index == 0)
+            {
+                return aliasQualifier == null ? (NameSyntax)namePiece : SyntaxFactory.AliasQualifiedName(aliasQualifier, namePiece);
+            }
+            else
+            {
+                return SyntaxFactory.QualifiedName(CreateNameSyntax(namespaceParts, aliasQualifier, index - 1), namePiece);
+            }                                                                              
+        }
+
+        protected override async Task<(bool, NamespaceDeclarationSyntax)> ShouldPositionTriggerRefactoringAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            
+            var namespaceDeclarationCount = root.DescendantNodes().OfType<NamespaceDeclarationSyntax>().Count();
+
+            if (namespaceDeclarationCount == 1)
+            {
+                var token = root.FindToken(position);
+                if (token.Kind() == SyntaxKind.EndOfFileToken)
+                {
+                    token = token.GetPreviousToken();
+                }
+                // Should trigger if cursor is on the name of only namespace declaration in this document.
+                var namespaceDeclaration = token.GetAncestor<NamespaceDeclarationSyntax>();
+
+                var shouldTrigger = namespaceDeclaration != null 
+                    && namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.GetDiagnostics().All(diag => diag.Severity != DiagnosticSeverity.Error)
+                    && !ContainsPartialDeclaration(document, namespaceDeclaration, cancellationToken);
+                return (shouldTrigger, shouldTrigger ? namespaceDeclaration : null);
+            }
+
+            if (namespaceDeclarationCount == 0 && root is CompilationUnitSyntax compilationUnit)
```

Comment:
Note... i'm not going to push back too heavily on this. However, this seems like a big corner case to have to handle.  Most people will have namespaces.  ALl our templates start you out in a namespace, and all new file templates generate a namespace...

so we should check if we really need to handle this.  It may simplify things a bunch.